tests/test_optimize.py

```diff
@@ -671,3 +718,149 @@ def test_update_stepsize(self):
+        assert x_new_one_wc != pytest.approx(x_new_two_wc, abs=tol)
+
+        if opt_name != "nest":
+            assert cost_one == pytest.approx(func(x, c=1.0), abs=tol)
+            assert cost_two == pytest.approx(func(x, c=2.0), abs=tol)
+
+    @pytest.mark.parametrize(
+        "func, args",
+        [
+            (lambda x, y: x * y, (1.0, 1.0)),
+            (lambda x, y: x[0] * y[0], (np.array([1.0]), np.array([1.0]))),
+        ],
+    )
+    def test_multi_args(self, opt, opt_name, func, args, tol):
+        """Test multiple arguments to function"""
+        x_new, y_new = opt.step(func, *args)
+        x_new2, y_new2 = opt.step(func, x_new, y_new)
+
+        (x_new_wc, y_new_wc), cost = opt.step_and_cost(func, *args)
+        (x_new2_wx, y_new2_wc), cost2 = opt.step_and_cost(func, x_new_wc, y_new_wc)
+
+        if getattr(opt, "reset", None):
+            opt.reset()
+
+        assert x_new != pytest.approx(args[0], abs=tol)
+        assert y_new != pytest.approx(args[1], abs=tol)
+
+        assert x_new_wc != pytest.approx(args[0], abs=tol)
+        assert y_new_wc != pytest.approx(args[1], abs=tol)
+
+        if opt_name != "nest":
+            assert cost == pytest.approx(func(*args), abs=tol)
+
+    def test_nontrainable_data(self, opt, opt_name, tol):
+        """Check non-trainable argument does not get updated"""
+
+        def func(x, data):
+            return x[0] * data[0]
+
+        x = np.array([1.0])
+        data = np.array([1.0], requires_grad=False)
+
+        args_new = opt.step(func, x, data)
+        args_new_wc, cost = opt.step_and_cost(func, *args_new)
+
+        if getattr(opt, "reset", None):
+            opt.reset()
+
+        assert len(args_new) == pytest.approx(2, abs=tol)
+        assert args_new[0] != pytest.approx(x, abs=tol)
+        assert args_new[1] == pytest.approx(data, abs=tol)
+
+        if opt_name != "nest":
+            assert cost == pytest.approx(func(args_new[0], data), abs=tol)
+
+    def test_multiargs_data_kwargs(self, opt, opt_name, tol):
+        """ Check all multiargs, non-trainable data, and keywords at the same time."""
+
+        def func(x, data, y, c=1.0):
+            return c * (x[0] + y[0] - data[0]) ** 2
+
+        x = np.array([1.0], requires_grad=True)
+        y = np.array([1.0])
+        data = np.array([1.0], requires_grad=False)
+
+        args_new, cost = opt.step_and_cost(func, x, data, y, c=0.5)
+        args_new2 = opt.step(func, *args_new, c=0.5)
+
+        if getattr(opt, "reset", None):
+            opt.reset()
+
+        assert args_new[0] != pytest.approx(x, abs=tol)
+        assert args_new[1] == pytest.approx(data, abs=tol)
+        assert args_new[2] != pytest.approx(y, abs=tol)
+
+        if opt_name != "nest":
+            assert cost == pytest.approx(func(x, data, y, c=0.5), abs=tol)
+
+    def test_steps_the_same(self, opt, opt_name, tol):
+        """Tests optimizing single parameter same as with several at a time"""
```

Comment:
I'm not quite sure I follow this docstring. I find the 'same as with several' slightly hard to parse?

Reply:
Could be good to expand this docstring to make the purpose of this test slightly clearer

Reply:
From other tests, we know that the step for a single argument is correct.  I want to make sure it has the same correct step regardless of position in the function call.

I will try to come up with a better way of explaining it concisely.